enhancements/ingress/add-headers-to-haproxyconfig-via-ingress-controller-or-route-api.md

```diff
@@ -0,0 +1,295 @@
+---
+title: add-headers-to-haproxyconfig-via-ingress-controller-or-route-api
+authors:
+  - "@miheer"
+reviewers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+approvers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+api-approvers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+creation-date: 2020-06-10
+last-updated: 2020-06-10
+tracking-link:
+  - "https://issues.redhat.com/browse/NE-982"
+status: implemented
+see-also:
+- ""
+replaces:
+superseded-by:
+---
+
+# Set HTTP Headers to haproxy.config either via Ingress Contoller or Route API
+
+## Summary
+
+This enhancement extends the IngressController logging API and router API to allow the user to set 
+capture of HTTP headers in the haproxy.config file. Using the new API, users may specify a list of 
+HTTP headers either via IngressController API or Route API that should be set in the haproxy.config file.
+
+## Motivation
+
+Certain headers may need to be set for compliance reasons or other reasons. In OpenShift 3, cluster administrators could
+use a custom HAProxy configuration template to set headers, but using a custom template is not permitted in OpenShift 4.
+
+### User Stories
+
+As a cluster administrator, I need to configure an IngressController to set certain HTTP headers,
+in order to satisfy compliance requirements.
+To satisfy this use-case, the cluster administrator can either set the IngressController's or Router's spec.httpSetHeaders
+field to set the required headers.
+
+### Goals
+
+1. Enable the cluster administrator to specify HTTP headers in the IngressController CR Spec that should get set in the global section
+   of the haproxy.config.
+2. Enable the developer or user to specify HTTP headers in the Route Spec that should get set in the backend section `default_backend openshift_default`.
+   of the haproxy.config.
+
+### Non-Goals
+
+1. Admins or users are responsible for providing proper header values as inputs.
+
+## Proposal
+
+### Workflow Description
+
+**Cluster admin** is a human user responsible for deploying a
+cluster.
+
+**Users/Devops/Developers** is a human user responsible for developing and
+deploying an application in a cluster.
+
+1. Cluster admin wants to set HTTP headers in the global section of the haproxy.config
+2. Cluster admin edits/creates the CR of the Ingress Controller having Spec as follows -
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  httpSetHeaders:
+  - name: X-Frame-Options
+    value: DENY
+  - name: X-XSS-Protection
+    value: 1;mode=block
```

Comment:
Where exactly the headers specified on the ingresscontroller level should be set in HAproxy config? Will they be set on all the existing routes (backends in HAProxy config)?

Reply:
What about the visibility for the user? Should these headers be added to the route's status or to the route's spec?

Reply:
spec